framework/auth/core.py

```diff
@@ -445,13 +445,16 @@ def create(cls, username, password, fullname):
         return user
 
     @classmethod
-    def create_unconfirmed(cls, username, password, fullname, do_confirm=True):
+    def create_unconfirmed(cls, username, password, fullname, do_confirm=True,
+                           campaign=None):
         """Create a new user who has begun registration but needs to verify
         their primary email address (username).
         """
         user = cls.create(username, password, fullname)
         user.add_unconfirmed_email(username)
         user.is_registered = False
+        if campaign:
+            user.system_tags.append(campaign)
```

Comment:
Is `system_tags` the best place to store campaign information? Since our mechanism of identifying campaign membership is based on just looking for a generic word in a catchall field, I'm worried that any other tag using the word "prereg" elsewhere later could accidentally trigger this functionality.

(it's a very common word- used 158 times in our codebase)

If we do continue using this field, here are two suggestions that could help:

1. In `VALID_CAMPAIGNS`, change `prereg` to `prereg_challenge_campaign` (make the special word more unique)
2. Building on the idea of `VALID_CAMPAIGNS`: Create a single central place where we can look up valid user system tags. Add a validator on the model field to ensure that only recognized tags can be used.

From recent experience with draft registrations, I can confirm that field validators help to avoid accidental mistakes or deletions of code, by helping other developers know what values can / can't be used in a field.